src/include/storage/recovery/replication_log_provider.h

```diff
@@ -1,20 +1,151 @@
 #pragma once
 
 #include <memory>
+#include <queue>
+#include <string>
+#include <utility>
+#include <vector>
 
+#include "network/network_io_utils.h"
 #include "storage/recovery/abstract_log_provider.h"
+#include "storage/write_ahead_log/log_io.h"
 
 namespace noisepage::storage {
 
 /**
- * TODO(Tianlei/Gus): Replace this file with Gus's version
+ * Log provider for logs being received over network.
+ * Provides logs to the recovery manager from logs being sent by master node over the network.
  */
-class ReplicationLogProvider {
+class ReplicationLogProvider final : public AbstractLogProvider {
  public:
   /**
-   * Passes the content of the buffer to traffic cop
-   * @param buffer content to pass to traffic cop
+   * Create a new ReplicationLogProvider.
+   *
+   * @param replication_timeout         The replication timeout. TODO(WAN): kill this?
    */
-  virtual void HandBufferToReplication(std::unique_ptr<network::ReadBuffer> buffer);
+  explicit ReplicationLogProvider(std::chrono::seconds replication_timeout)
+      : replication_active_(true), replication_timeout_(replication_timeout) {}
+
+  LogProviderType GetType() const override { return LogProviderType::REPLICATION; }
+
+  /**
+   * Notifies the log provider that replication is ending.
+   */
+  void EndReplication() {
+    std::unique_lock<std::mutex> lock(replication_latch_);
+    replication_active_ = false;
+    replication_cv_.notify_all();
+  }
+
+  /** Buffer the message. */
+  void AddBufferFromMessage(const uint64_t primary_callback_id, const std::string &content) {
+    std::vector<unsigned char> bytes(content.begin(), content.end());
+    network::ReadBufferView view(bytes.size(), bytes.begin());
+    auto buffer = std::make_unique<network::ReadBuffer>();
+    buffer->FillBufferFrom(view, bytes.size());
+    {
+      std::unique_lock<std::mutex> lock(replication_latch_);
+      arrived_buffer_queue_.emplace(primary_callback_id, std::move(buffer));
+      replication_cv_.notify_one();
+    }
+
+    // TODO(WAN): if you want sync/async I think this is where you do it
+  }
+
+  // TODO(gus): this method is not entirely correct, there could still be messages over network
+  /**
+   * TODO(WAN): Per Gus, tThis method is not entirely correct because there could still be messages pending over
+   *  the network.
+   */
+  void WaitUntilSync() {
+    while ((curr_buffer_ != nullptr && curr_buffer_->HasMore()) || !arrived_buffer_queue_.empty())
```

Comment:
I guess it doesn't really matter since this isn't correct but I think you could just replace this with
`while(NonBlockingHasMoreRecords()) {`

Reply:
It was actually never used, I just removed it.